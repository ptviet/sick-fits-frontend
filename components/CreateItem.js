import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import gql from 'graphql-tag';
import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $image: String
    $largeImage: String
    $price: Int!
  ) {
    createItem(
      title: $title
      description: $description
      image: $image
      largeImage: $largeImage
      price: $price
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0
  };

  onChange = e => {
    const { name, type, value } = e.target;
    const val = type === 'number' ? parseFloat(Number(value)) : value;
    this.setState({ [name]: val });
  };

  data = null;
  onFileChange = e => {
    e.preventDefault();
    const files = e.target.files;
    this.data = new FormData();
    this.data.append('file', files[0]);
    this.data.append('upload_preset', 'sickfits');
    this.previewImage(files[0]);
  };

  previewImage(file) {
    if (file != null || file != undefined) {
      var oFReader = new FileReader();
      oFReader.readAsDataURL(file);

      oFReader.onload = oFREvent => {
        document.getElementById('uploadPreview').src = oFREvent.target.result;
        document.getElementById('uploadPreview').alt = 'Upload Preview';
        document.getElementById('uploadPreview').style = 'height: 100px';
        document.getElementById('uploadPreview').hidden = false;
      };
    } else {
      document.getElementById('uploadPreview').hidden = true;
      this.data = null;
    }
  }

  onSubmit = async (e, createItemMutation) => {
    e.preventDefault();
    if (this.data != null) {
      const uploadRes = await fetch(
        'https://api.cloudinary.com/v1_1/ptviet/image/upload',
        {
          method: 'POST',
          body: this.data
        }
      );
      const file = await uploadRes.json();
      await this.setState({
        image: file.secure_url,
        largeImage: file.eager[0].secure_url
      });
      const res = await createItemMutation();
      Router.push({
        pathname: '/item',
        query: { id: res.data.createItem.id }
      });
    }
  };

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form onSubmit={e => this.onSubmit(e, createItem)}>
            <ErrorMessage error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <img id="uploadPreview" />
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.onFileChange}
                />
              </label>
              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  value={this.state.title}
                  required
                  onChange={this.onChange}
                />
              </label>
              <label htmlFor="price">
                Price (in cent)
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  value={this.state.price}
                  required
                  onChange={this.onChange}
                />
              </label>
              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Description"
                  value={this.state.description}
                  required
                  onChange={this.onChange}
                />
              </label>
              <button type="submit">Submit{loading && 'ting'}</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
