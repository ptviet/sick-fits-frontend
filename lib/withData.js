import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpointProd } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    // uri: process.env.NODE_ENV === 'development' ? endpoint : endpointProd,
    uri: 'https://sp-sick-fits-yoga-prod.herokuapp.com',
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include'
        },
        headers
      });
    },
    // Local data
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            // Read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({ query: LOCAL_STATE_QUERY });
            // Write new cartOpen state
            const data = {
              data: {
                cartOpen: !cartOpen
              }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
