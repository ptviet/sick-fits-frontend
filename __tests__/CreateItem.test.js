import { MockedProvider } from 'react-apollo/test-utils';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import wait from 'waait';
import Router from 'next/router';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { fakeItem } from '../lib/testUtils';

Router.router = { push: jest.fn() };

const item = fakeItem();

// Mock the global fetch API
// global.fetch = jest.fn().mockResolvedValue({
//   json: () => ({
//     secure_url: item.image,
//     eager: [{ secure_url: item.image }]
//   })
// });

const mocks = [
  {
    // when
    request: {
      query: CREATE_ITEM_MUTATION,
      variables: {
        title: item.title,
        description: item.description,
        image: '',
        largeImage: '',
        price: item.price
      }
    },
    // then
    result: {
      data: {
        createItem: {
          id: item.id,
          __typename: 'Item'
        }
      }
    }
  }
];

const type = (wrapper, input, name, value) => {
  const element = wrapper.find(`${input}[name="${name}"]`);
  element.prop('onChange')({ target: { name, value } });
};

describe('<CreateItem />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>
    );
  });
  it('should render and match the snapshot', () => {
    const form = wrapper.find('form');
    expect(toJson(form)).toMatchSnapshot();
  });

  it('should preview the image', async () => {
    const component = wrapper.find('CreateItem').instance();
    expect(component.data).toBeNull();

    const uploadPreview = wrapper.find('#uploadPreview');
    expect(uploadPreview.props().src).toBeUndefined();

    // Mock uploadPreview
    const file = new Blob(['dog.jpg'], { type: 'image/jpg' });
    const readAsDataURL = () => {
      uploadPreview.props().src = item.image;
    };

    const dummyFileReader = {
      readAsDataURL
    };
    const fileUpload = wrapper.find('input[type="file"]');
    window.FileReader = jest.fn(() => dummyFileReader);
    fileUpload.simulate('change', { target: { files: [file] } });

    await wait();
    wrapper.update();

    expect(component.data).not.toBeNull();
    expect(uploadPreview.props().src).toEqual(item.image);
  });

  it('should create an item when the form is submitted', async () => {
    // Simulate input
    type(wrapper, 'input', 'title', item.title);
    type(wrapper, 'textarea', 'description', item.description);
    type(wrapper, 'input', 'price', item.price);

    // component.setState({ image: item.image, largeImage: item.largeImage });

    // expect(component.state).toMatchObject({
    //   title: item.title,
    //   description: item.description,
    //   image: item.image,
    //   largeImage: item.largeImage,
    //   price: item.price
    // });

    // Submit the form
    const form = wrapper.find('form');
    form.simulate('submit');
    await wait(50);
    expect(Router.router.push).toHaveBeenCalled();

    // expect(global.fetch).toHaveBeenCalled();
    // global.fetch.mockreset();
  });
});
