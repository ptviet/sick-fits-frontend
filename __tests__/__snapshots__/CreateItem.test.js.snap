// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> should render and match the snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 fCpNnG"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <img
      alt="Upload Preview"
      id="uploadPreview"
    />
    <label
      htmlFor="file"
    >
      Image
      <input
        id="file"
        name="file"
        onChange={[Function]}
        placeholder="Upload an image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="price"
    >
      Price (in cent)
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      Description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
